### ISSUE TITLE
tf.train.batch and shuffle_batch undetermined for multi queue input with multiple threads

### ISSUE THREAD

**Issue by BastiaanBergman** [tensorflow/tensorflow#13847]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
Custom code, see below.

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
OSX

- **TensorFlow installed from (source or binary)**:
pip install tensorflow == 1.2

- **TensorFlow version (use command below)**:
('v1.2.0-rc2-21-g12f033d', '1.2.0')

- **Python version**: 
Python 2.7.10

- **Bazel version (if compiling from source)**:
n/a

- **CUDA/cuDNN version**:
n/a

- **GPU model and memory**:
n/a

- **Exact command to reproduce**:

```
a_fifo = tf.train.slice_input_producer([np.arange(30)],
                                        shuffle=False,
                                        name='a')[0]
b_fifo = tf.train.slice_input_producer([np.arange(30)],
                                        shuffle=False,
                                        name='b')[0]
train_batch = tf.train.batch([a_fifo, b_fifo],
                            batch_size=5,
                            capacity=18,
                            num_threads=2,
                            name='batch')
at, bt = train_batch
with tf.Session() as sess:
    threads = tf.train.start_queue_runners(sess=sess)
    sess.run(tf.global_variables_initializer())
    try:
        for i in range(150):
            av, bv = sess.run([at, bt])
            assert list(av) == list(bv), "queues derailed!"
    except Exception as e:
        print(("Exception in training: {}").format(e))
        print("__",i,"__")
        print(av)
        print(bv)
```

Output:

> Exception in training: queues derailed!
> __ 11 __
> [26 25 27 28 29]
> [25 26 27 28 29]

### Describe the problem
Both queues should be dequeued synchronously, as it does when num_threads =1 in the batch statement. With num_threads > 1 the two input queue's will derail, as shown in the example above. Same issue appears also in train.shuffle_batch.

### Source code / logs
See code above.

**Comment by drpngx** [comment]
@BastiaanBergman we are moving to `tf.data`, which is a much nicer interface. Would you be open to that?

**Comment by BastiaanBergman** [comment]
I really liked tf.data but found it not working for some other reason, therefore reverted back to the queues. Will retry some day.

**Comment by drpngx** [comment]
We are deprecating the queues in favor of `tf.data` and would really like to hear about what the problem is.

@mrry maybe we can suggest a solution?

**Comment by mrry** [comment]
Yes, the difficulty of getting deterministic results is one of the reasons why we are recommending that people use `tf.data` instead of the queue-based APIs. For example, here's what your program would look like in `tf.data`:

```python
a_dataset = tf.data.Dataset.range(30).repeat()
b_dataset = tf.data.Dataset.range(30).repeat()
batched_dataset = tf.data.Dataset.zip((a_dataset, b_dataset)).batch(5)
iterator = batched_dataset.make_one_shot_iterator()
at, bt = iterator.get_next()

with tf.Session() as sess:
    try:
        for i in range(150):
            av, bv = sess.run([at, bt])
            assert list(av) == list(bv), "queues derailed!"
    except Exception as e:
        print(("Exception in training: {}").format(e))
        print("__",i,"__")
        print(av)
        print(bv)
```

**Comment by BastiaanBergman** [comment]
Yes its nice that way. I still have to look and see what made me move away from dataset. I think it was issues with switching between test and train sets on the same graph. But I had issues with that in queue's too and eventually settled with separate graphs for training and testing anyway.

Are there multiple threads working on this dataset version? Would I be save with one thread on queue's? Or might I still have compromised 'deterministic results'? ;-) Thanks anyway, it's been awesome learning tf.

**Comment by mrry** [comment]
There are a couple of ways to switch between different datasets with a single `Iterator` (search for "reinitializable iterator" and "feedable iterator" in [this guide](https://www.tensorflow.org/programmers_guide/datasets#creating_an_iterator)), but most people find it easier to construct multiple graphs for these cases.

The code example I showed does not create additional threads, but you can add them wherever you'd like by adding a `.prefetch(buffer_size)` transformation. Although it doesn't make much sense for the concrete example, here's how you'd add prefetching threads to the `a_dataset` and `b_dataset` computations:

```python
a_dataset = tf.data.Dataset.range(30).repeat().prefetch(10)
b_dataset = tf.data.Dataset.range(30).repeat().prefetch(10)
batched_dataset = tf.data.Dataset.zip((a_dataset, b_dataset)).batch(5)
```

If you have an expensive `Dataset.map()` in your program, you can also parallelize that by adding `num_parallel_calls=N` to process `N` elements at once.

All these options add parallelism without comprising on determinism. There is an explicitly "sloppy" transformation called `tf.contrib.data.sloppy_interleave()` that can produce non-determinstic results (but can be useful to make I/O blockages), and we've considered adding a non-deterministic parallel map (to deal with stragglers), but it's generally been more pleasant to work with deterministic APIs.

By contrast, the `num_threads > 1` versions of the queue-based APIs are inherently racy, because they issue multiple `Session.run()` calls in parallel, and mutate queue state (dequeuing from the source queue, enqueuing to the destination queue) without any coordination. Setting `num_threads = 1` ought to help, but it conflicts with `tf.train.shuffle_batch()`, because there can still be a race between a single producer thread and a single consumer thread when `capacity != min_after_dequeue`, leading to unrepeatable results.

**Comment by mrry** [comment]
Closing due to lack of activity.
